### ISSUE TITLE
Standardize process by which each language performs normalization and extension management

### ISSUE THREAD

**Issue by matthchr** [Azure/autorest#1148]
Today, there is a single method: `public override void NormalizeClientModel(ServiceClient serviceClient)` which each code generator implements.

What exactly is done in this method is left entirely up to the code generator (as it should be)... but for many AutoRest languages, they are processing/supporting 90%+ the same set of extensions. The problem is that each language does this process in a somewhat different and complicated manner:

Compare **AzureCSharp:**
```
        public override void NormalizeClientModel(ServiceClient serviceClient)
        {
            AzureExtensions.NormalizeAzureClientModel(serviceClient, Settings, _namer);
            _namer.NormalizeClientModel(serviceClient);
            _namer.ResolveNameCollisions(serviceClient, Settings.Namespace,
                Settings.Namespace + ".Models");
            _namer.NormalizePaginatedMethods(serviceClient, pageClasses);
            _namer.NormalizeODataMethods(serviceClient);

            if (serviceClient != null)
            {
                foreach (var model in serviceClient.ModelTypes)
                {
                    if (model.Extensions.ContainsKey(AzureExtensions.AzureResourceExtension) && 
                        (bool)model.Extensions[AzureExtensions.AzureResourceExtension])
                    {
                        model.BaseModelType = new CompositeType { Name = "IResource", SerializedName = "IResource" };
                    }
                }
            }
        }
```

to **NodeJS:**
```
        public override void NormalizeClientModel(ServiceClient serviceClient)
        {
            // MethodNames are normalized explicitly to provide a consitent method name while 
            // generating cloned methods for long running operations with reserved words. For
            // example - beginDeleteMethod() insteadof beginDelete() as delete is a reserved word.
            Namer.NormalizeMethodNames(serviceClient);
            AzureExtensions.NormalizeAzureClientModel(serviceClient, Settings, Namer);
            base.NormalizeClientModel(serviceClient);
            NormalizeApiVersion(serviceClient);
            NormalizePaginatedMethods(serviceClient);
            ExtendAllResourcesToBaseResource(serviceClient);
        }
```

to **Python:**
```
        public override void NormalizeClientModel(ServiceClient serviceClient)
        {
            // Don't add pagable/longrunning method since we already handle ourself.
            Settings.AddCredentials = true;
            AzureExtensions.ProcessClientRequestIdExtension(serviceClient);
            AzureExtensions.UpdateHeadMethods(serviceClient);
            AzureExtensions.ParseODataExtension(serviceClient);
            Extensions.FlattenModels(serviceClient);
            ParameterGroupExtensionHelper.AddParameterGroups(serviceClient);
            AzureExtensions.AddAzureProperties(serviceClient);
            AzureExtensions.SetDefaultResponses(serviceClient);
            CorrectFilterParameters(serviceClient);

            base.NormalizeClientModel(serviceClient);
            NormalizeApiVersion(serviceClient);
            NormalizePaginatedMethods(serviceClient);
        }
```

(note: Java looks like CSharp basically, which is good!)

So you can see that 3 different languages  have 3 very different looking implementations of this method (and Ruby's is different too, I just didn't want to include yet another example):
* NodeJS and Python call the `base.NormalizeClientModel` but CSharp doesn't.
* CSharp and NodeJS use the helper method `AzureExtensions.NormalizeAzureClientModel` but Python doesn't.
* CSharp uses `_namer.NormalizePaginatedMethods(serviceClient)` whereas NodeJS and Python use `NormalizePaginatedMethods(serviceClient)`.
* CSharp does `AzureExtensions.NormalizeAzureClientModel(serviceClient, Settings, _namer)` immediately followed by `_namer.NormalizeClientModel(serviceClient)` -- it's unclear (at least to me) what each of those different calls do and why one needs to happen before the other (until you look at the implementation, then it's clear).

This is a bit of a catchall issue, but **a few concrete improvements would be**:
1. Clarify what the `namer` is for, and use it for those things across all the languages.
2. Make the method structures look a bit more similar (see CSharp + Java for a good example) -- for example things like all Azure generators calling their `base.NormalizeClientModel` (or all of them not).
3. Enable the `AzureExtensions.NormalizeAzureClientModel` method to take a parameter which allows you to opt out of particular bits, so that at least order is enforced even if not all of the extensions are applied (this would allow Python to use it instead of rolling its own I think, and maybe Ruby too).
4. Maybe rename `AzureExtensions.NormalizeAzureClientModel` to something like `AzureExtensions.ApplyAzureClientExtensions` since that's actually what it's doing (I think).

**Comment by fearthecowboy** [comment]
I noticed this too, and was kinda curious about it.

I think this deserves some discussion, and I'll include in an upcoming meeting where we're going to look at grouping some issues into categories and set some direction on addressing them.

G

**Comment by matthchr** [comment]
@fearthecowboy This was actually discussed in Thursday's team sync up meeting (not sure if you were there).
@tbombach might be able to clarify exactly what planned actions were going to be taken, but I got the impression that it was:
* Link/group this item with other items related to extension management (i.e. the feature where each code generator can report what extensions it supports) -- when making that improvement this could be done as well.
* Maybe make some intermediate changes to address my comments (1-4) in the interim if it will be a while before the bigger extension management feature is done.

**Comment by fearthecowboy** [comment]
No, I wasn't there, I was out sick. :no_good: 

I have some larger planning themes that I'm preparing, I'll likely be sharing them at the next meeting.

G

**Comment by fearthecowboy** [comment]
Howdy!

In our planning for driving towards a stable '1.0' release, I'm marking this issue as 'deferred' :zzz: and we're going to review it during the post-1.0 planning cycle. 

_It's not to say that we're not going to work on it, or that this isn't not important, but at the moment, we're picking and choosing the stuff we **must** do before 1.0._ :horse_racing:  :horse_racing:  :horse_racing: 

We'll make sure we pick this back up at that point. :tada:
